rs/parser/src/nodes.rs

```diff
@@ -0,0 +1,924 @@
+
+impl_single_item_classes!(
+    (
+        Delayn,
+    ) => param: UsizeOrRef<'ast>,
+    (
+        Delayms,
+        Imp,
+        Tri,
+        Squ,
+        Saw,
+        Onepole,
+        Sin,
+        Mul,
+        Add,
+        Pan,
+        Bd,
+        Sn,
+        Hh,
+    ) => param: NumberOrRef<&'ast str>,
+    (
+        Meta,
+        Expr,
+        Eval,
+    ) => code: CodeBlock<'ast>,
+);
+
+#[derive(PartialEq, Debug)]
+pub struct Speed {
+    pub speed: f32
+}
+
+impl Node<'_> for Speed {
+    fn parse_from_iter(pairs: &mut Pairs<'_, Rule>, span: Span<'_>) -> Result<Self, Box<Error<Rule>>> {
+        pairs.next_parsed(span)
+            .map(|speed| Self { speed })
+    }
+}
+
+#[derive(PartialEq, Debug)]
+pub struct Get<'ast> {
+    pub reference: &'ast str
+}
+
+impl<'ast> Node<'ast> for Get<'ast> {
+    #[cfg_attr(test, trace::trace(prefix_enter = "[+ Get]"))]
+    fn parse_from_iter(pairs: &mut Pairs<'ast, Rule>, span: Span<'ast>) -> Result<Self, Box<Error<Rule>>> {
+        pairs.next()
+            .ok_or_else(|| span.as_end_span().to_err_with_positives([Rule::reference]))
+            .map(|p| Self { reference: p.as_str() })
+    }
+}
+
+#[derive(PartialEq, Debug)]
+pub struct Noise {
+    pub seed: usize
+}
+
+impl Node<'_> for Noise {
+    #[cfg_attr(test, trace::trace(prefix_enter = "[+ Noise]"))]
+    fn parse_from_iter(pairs: &mut Pairs<'_, Rule>, span: Span<'_>) -> Result<Self, Box<Error<Rule>>> {
+        pairs.next_parsed(span).map(|seed| Self { seed })
+    }
+}
+
+#[derive(PartialEq, Debug)]
+pub enum UsizeOrRef<'ast> {
+    Usize(usize),
+    Ref(&'ast str)
+}
+
+impl<'ast> Node<'ast> for UsizeOrRef<'ast>{
+    #[cfg_attr(test, trace::trace(prefix_enter = "[+ UsizeOrRef]"))]
+    fn parse_from_iter(pairs: &mut Pairs<'ast, Rule>, span: Span<'ast>) -> Result<Self, Box<Error<Rule>>> {
+        let next = pairs.next()
+            .ok_or_else(|| span.as_end_span().to_err_with_positives([Rule::integer, Rule::reference]))?;
+
+        match_or_return_err!(next,
+            Rule::number => {
+                next.try_to_parse().map(Self::Usize)
+            },
+            Rule::reference => {
+                Ok(Self::Ref(next.as_str()))
+            },
+        )
+    }
+}
+
+#[derive(PartialEq, Debug)]
+pub struct Adc {
+    pub port: u32
+}
+
+impl Node<'_> for Adc {
+    #[cfg_attr(test, trace::trace(prefix_enter = "[+ Adc]"))]
+    fn parse_from_iter(pairs: &mut Pairs<'_, Rule>, span: Span<'_>) -> Result<Self, Box<Error<Rule>>> {
+        pairs.next_parsed::<u32>(span)
+            .map(|port| Self { port })
+    }
+}
+
+#[derive(PartialEq, Debug)]
+pub struct Plate {
+    pub mix: f32
+}
+
+impl Node<'_> for Plate {
+    #[cfg_attr(test, trace::trace(prefix_enter = "[+ Plate]"))]
+    fn parse_from_iter(pairs: &mut Pairs<'_, Rule>, span: Span<'_>) -> Result<Self, Box<Error<Rule>>> {
+        pairs.next_parsed(span).map(|mix| Self { mix })
+    }
+}
+
+// TODO: There are some ambiguous 'seq' inputs, afaict.
```

Comment:
try https://pest.rs/#editor and paste the glicol.pest

`seq o_` or `seq _o` should not compile.

`seq` takes ref with tilde (maybe we should call it control sig ref in the future?) such as `~a`, `60`(note) or `_` (rest).

so `seq ~o_` will play ~o at bar portion 0.0, and play a "rest" at bar portion 0.5. the latter will be automatically ignored of course. 

`seq _~o` will play ~o at 0.5.

`seq 6060` is invalid musically as 6060 will be seen as a (extremely high) midi note...

use `seq 60_60_` or simply write `seq 60 60`

same for `seq ~a_14_` 

the idea is always to divide one bar with spaces first; the divided part is called `compound note`, which is a permutation of tilde ref, note number and rest. then we further divide the compound note based on the amount of the element inside

hope this answers your question